### PR TITLE
Set scheme for traefik_route

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
Grafana, which uses traefik_route, is currently unable to tell if the ingress endpoint is http or https.

## Solution
This PR adds a scheme arg, so that traefik could inform over relation data which scheme it is using.

In tandem with:
- https://github.com/canonical/traefik-route-k8s-operator/pull/41
- https://github.com/canonical/grafana-k8s-operator/pull/257

Drive-by fixes:
- [IPU] Abort provide_ingress_requirements if there is no relation. The same thing was already impl'd for IPA in #233.

## Release Notes
Set scheme for traefik_route.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -85,7 +85,7 @@ allowlist_externals = /usr/bin/env
 description = Scenario tests
 deps =
     pytest
-    ops-scenario>=4.0.4
+    ops-scenario<5
```
Comment (sed-i) [inline tox.ini:88]: @PietroPasotti I tried `>=5`, unpinned ops and ran `tox -re`, but there were `_fs` errors and others.

File: lib/charms/traefik_route_k8s/v0/traefik_route.py [hunk 2]
```diff
@@ -189,30 +200,43 @@ def _update_stored_external_host(self) -> None:
...
+            return
+
+        for relation in self._charm.model.relations[self._relation_name]:
+            if not relation.app:
+                self._stored.external_host = ""
+                self._stored.scheme = ""
+                return
+            external_host = relation.data[relation.app].get("external_host", "")
+            self._stored.external_host = external_host or self._stored.external_host  # type: ignore
+            scheme = relation.data[relation.app].get("scheme", "")
```
Comment (Abuelodelanada) [inline lib/charms/traefik_route_k8s/v0/traefik_route.py:213]: Shouldn't we be explicit about the scheme here?

```suggestion
            scheme = relation.data[relation.app].get("scheme", "http")
```

**Review by Abuelodelanada — APPROVED** [review]

File: lib/charms/traefik_route_k8s/v0/traefik_route.py (hunk 2 above)
Comment (sed-i) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:213]: I think you're right, but I was following the existing pattern for `external_host`.
There are more odd things here that could use a follow-up PR:
- `_stored` gets overwritten every loop iteration.
- Why the provider needs to extract this from relation data? Seems like this was cloned from the requirer. Not sure.

Comment (sed-i) [reply lib/charms/traefik_route_k8s/v0/traefik_route.py:213]: https://github.com/canonical/traefik-route-k8s-operator/issues/42

**Pull request merged**